src/Evolution/Systems/Cce/ScriPlusInterpolationManager.hpp

```diff
@@ -0,0 +1,607 @@
+  /// but for full accuracy, check `first_time_is_ready_to_interpolate` before
+  /// calling the interpolation functions
+  std::pair<double, ToInterpolate> interpolate_peek_first_time() noexcept;
+
+  /// \brief Interpolate to the first targe time in the queue, returning both
+  /// the time and the interpolated data at that time, and remove the first time
+  /// from the queue.
+  ///
+  /// note: This will work regardless of whether it is 'ready' to interpolate
+  /// (has the target number of points before and after). This functionality is
+  /// provided to allow obtaining rough answers before the data is fully ready,
+  /// but for full accuracy, check `first_time_is_ready_to_interpolate` before
+  /// calling the interpolation functions
+  std::pair<double, ToInterpolate> interpolate_and_pop_first_time() noexcept;
+
+  /// return the number of times in the target times queue
+  size_t target_times_size() noexcept { return target_times_.size(); }
+
+  /// return the number of data points that have been provided to the
+  /// interpolation manager
+  size_t data_sizes() noexcept { return u_bondi_ranges_.size(); }
+
+ private:
+  void cull_unneeded_early_times() noexcept;
+
+  friend struct ScriPlusInterpolationManager<ToInterpolate, Tags::Du<Tag>>;
+
+  std::deque<DataVector> u_bondi_values_;
+  std::deque<ToInterpolate> to_interpolate_values_;
+  std::deque<std::pair<double, double>> u_bondi_ranges_;
+  std::deque<double> target_times_;
+  size_t vector_size_;
+  size_t target_number_of_points_;
+  std::unique_ptr<intrp::SpanInterpolator> interpolator_;
+};
+
+template <typename ToInterpolate, typename Tag>
+bool ScriPlusInterpolationManager<
+    ToInterpolate, Tag>::first_time_is_ready_to_interpolate() const noexcept {
+  size_t maxes_below = 0;
+  size_t mins_above = 0;
+  for (auto a : u_bondi_ranges_) {
+    if (a.first >= target_times_.front()) {
+      ++mins_above;
+    }
+    if (a.second <= target_times_.front()) {
+      ++maxes_below;
+    }
+  }
+  // we might ask for a time that's too close to the end or the beginning of
+  // our data, in which case we will settle for at least one point below and
+  // above and a sufficient number of total points.
+  // This class will never be 'ready' to extrapolate.
+  return ((maxes_below > target_number_of_points_) and
+          (mins_above > target_number_of_points_)) or
+         (maxes_below + mins_above > 2 * target_number_of_points_ and
+          maxes_below > 0 and mins_above > 0);
+}
+
+template <typename ToInterpolate, typename Tag>
+std::pair<double, ToInterpolate> ScriPlusInterpolationManager<
+    ToInterpolate, Tag>::interpolate_peek_first_time() noexcept {
+  if (target_times_.empty() or
+      to_interpolate_values_.size() < 2 * target_number_of_points_) {
+    return std::make_pair(0.0, ToInterpolate{});
+  }
+  // note that because we demand at least a certain number before and at least
+  // a certain number after, we are likely to have a surfeit of points for the
```

Comment:
we are not cultured enough to know surfeit...

Reply:
turns out the note isn't all that relevant for the updated method anyway, so I'll just remove it.